File: 2015/7.jl
function part1()
    f = open("instructions7a.txt", "r")
    lines = readlines(f)

    vals = Dict{String, Int32}()
    isintstring(str) = all(isdigit(c) for c in str)

    while length(lines) != 0
        new_lines = String[]    
        for (il, l) in enumerate(lines)
            cmds = split(l, "->")
            lcmds = split(cmds[1])
            rcmds = split(cmds[2])

            #println(rcmds[1])
            try
                if length(lcmds) == 1
                    if isintstring(lcmds[1])
                        vals[rcmds[1]] = parse(Int32, lcmds[1])
                    else
                        vals[rcmds[1]] = vals[lcmds[1]]    
                    end
                elseif length(lcmds) == 2
                    if lcmds[1] == "NOT"
                        vals[rcmds[1]] = 65535 - vals[lcmds[2]]
                    end
                elseif length(lcmds) == 3
                    if lcmds[2] == "AND"
                        if isintstring(lcmds[1])
                            vals[rcmds[1]] = parse(Int32, lcmds[1]) & vals[lcmds[3]]
                        else
                            vals[rcmds[1]] = vals[lcmds[1]] & vals[lcmds[3]] 
                        end
                    elseif lcmds[2] == "OR"
                        if isintstring(lcmds[1])
                            vals[rcmds[1]] = parse(Int32, lcmds[1]) | vals[lcmds[3]]
                        else
                            vals[rcmds[1]] = vals[lcmds[1]] | vals[lcmds[3]]
                        end
                    elseif lcmds[2] == "LSHIFT"
                        vals[rcmds[1]] = vals[lcmds[1]] << parse(Int32, lcmds[3])
                    elseif lcmds[2] == "RSHIFT"
                        vals[rcmds[1]] = vals[lcmds[1]] >> parse(Int32, lcmds[3])
                    end
                end
            catch e
                push!(new_lines, l)
            end
        end
        lines = new_lines
    end
    println("🎅 ", vals["a"])
end

function part2()
    f = open("instructions7b.txt", "r")
    lines = readlines(f)

    vals = Dict{String, Int32}()
    isintstring(str) = all(isdigit(c) for c in str)

    while length(lines) != 0
        new_lines = String[]    
        for (il, l) in enumerate(lines)
            cmds = split(l, "->")
            lcmds = split(cmds[1])
            rcmds = split(cmds[2])

            #println(rcmds[1])
            try
                if length(lcmds) == 1
                    if isintstring(lcmds[1])
                        vals[rcmds[1]] = parse(Int32, lcmds[1])
                    else
                        vals[rcmds[1]] = vals[lcmds[1]]    
                    end
                elseif length(lcmds) == 2
                    if lcmds[1] == "NOT"
                        vals[rcmds[1]] = 65535 - vals[lcmds[2]]
                    end
                elseif length(lcmds) == 3
                    if lcmds[2] == "AND"
                        if isintstring(lcmds[1])
                            vals[rcmds[1]] = parse(Int32, lcmds[1]) & vals[lcmds[3]]
                        else
                            vals[rcmds[1]] = vals[lcmds[1]] & vals[lcmds[3]] 
                        end
                    elseif lcmds[2] == "OR"
                        if isintstring(lcmds[1])
                            vals[rcmds[1]] = parse(Int32, lcmds[1]) | vals[lcmds[3]]
                        else
                            vals[rcmds[1]] = vals[lcmds[1]] | vals[lcmds[3]]
                        end
                    elseif lcmds[2] == "LSHIFT"
                        vals[rcmds[1]] = vals[lcmds[1]] << parse(Int32, lcmds[3])
                    elseif lcmds[2] == "RSHIFT"
                        vals[rcmds[1]] = vals[lcmds[1]] >> parse(Int32, lcmds[3])
                    end
                end
            catch e
                push!(new_lines, l)
            end
        end
        lines = new_lines
    end
    println("🎅 ", vals["a"])
end

part1()
part2()

File: 2015/7.py
import time, re

from collections import deque

from utils import readInput

r = re.compile("[a-z]{1,4}")

def loadInput():
    lines = readInput("instructions7a.txt")
    return deque(lines)

def decrypt(m):
    m = m.split("->")
    matches = r.findall(m[0])
    m[-1] = m[-1].strip()

    for v in matches:
        m[0]= m[0].replace(v, "____['" + v + "']")
    cmd = m[-1] + "=" + m[0].replace("____", "vars")

    cmd = cmd.replace("RSHIFT", ">>")
    cmd = cmd.replace("LSHIFT", "<<")
    cmd = cmd.replace("NOT", "65535-")
    cmd = cmd.replace("AND", "&")
    cmd = cmd.replace("OR", "|")
    return cmd

def part1(instructions):
    vars = {}
    while len(instructions) != 0:
        m = instructions.popleft()
        if m.startswith("#"):
            continue
        cmd = decrypt(m)
        try:
            #print("executing:", cmd)
            c = cmd.split("=")
            vars[c[0]] = eval(c[1])
        except:
            instructions.append(m)
    print (f"🎄 Part 1: {vars["a"]}")
    return vars['a']

def reset(instructions, a):
    for i in range(len(instructions)):
        if instructions[i].endswith(" -> b"):
            instructions[i] = f"{a} -> b"

def part2(instructions):
    vars = {}
    while len(instructions) != 0:
        m = instructions.popleft()
        cmd = decrypt(m)
        try:
            #print("executing:", cmd)
            c = cmd.split("=")
            vars[c[0]] = eval(c[1])
        except:
            instructions.append(m)
    print (f"🎄🎅 Part 2: {vars["a"]}")

if __name__ == "__main__":
    title = "Day 7: Some Assembly Required"
    sub = "-"*(len(title)+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput()

    t0 = time.time()
    a = part1(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
    
    inputs = loadInput()
    reset(inputs, a)

    t0 = time.time()
    part2(inputs)
    print ("Time: {:.5f}".format(time.time()-t0))
